### ISSUE TITLE
Make KPXC databases accessible via libsecret/DBus

### ISSUE THREAD

**Issue by kwill** [keepassxreboot/keepassxc#1403]
**User story:** I am a GNU/Linux desktop user with a standard (libsecret-based) password manager. [1] I want to use the same password manager to access my KPXC databases.

[1] For example using [GNOME Passwords and Keys](https://help.gnome.org/users/seahorse/stable/) UI (aka Seahorse) to access [GNOME Keyring](https://wiki.gnome.org/Projects/GnomeKeyring/) daemon.

## Possible Solution
* Make a libsecret-KPXC interface
* Test that it accessible from GNOME Keyring (for example)

**Comment by kwill** [comment]
It's what I'm thinking of. If you can provide better or alternate examples, or if I've been unclear, please suggest them and I'll modify the description.

**Comment by eppelmania** [comment]
I think one intention of issue #440 was to use KPXC as system keyring. So there would be no more need for seahorse or kwallet.

The user story would be more like the following:
I am a GNU/Linux desktop user and I want to use keepassxc instead of seahorse/kwallet.

**Comment by kabili207** [comment]
Seahorse and kwallet are user interfaces that interact with a libsecret keyring, they are not the actual keyring. The actual underlying keyring implementations are called gnome-keyring and ksecretservice.

This user story is about replacing the underlying implementations. Anything that would normally use gnome-keyring, such as seahorse or chrome or whatever, will talk to keepassxc instead.

**Comment by eppelmania** [comment]
Ok. So the possible solution would be. Make a libsecret-KPXC secret und test if it is accessible from seahorse, chrome, etc.

But shouldn't there also be some session handling stuff involved, for example automatic unlocking of the database after login, or is this already part of the libsecret interface.

**Comment by kabili207** [comment]
Libscret does include stuff in the API that signals the keyring to open a "master password" entry screen among other things, but I think that a libsecret-kpxc daemon would need to be running in the background for *any* of the API to even work.

**Comment by kakra** [comment]
@kabili207 As far as I understood, the secret service API is dbus activated. So "the daemon" is already there and could start keepass on demand.

**Comment by kabili207** [comment]
@kakra You're probably right. It's been quite some time since I've looked into the secret service API.

**Comment by mjsir911** [comment]
Would this launch as a very minimal application just acting as a daemon to serve as a libsecret password server or would the secret service API start up the entire KPXC program, GUI and all?

**Comment by kakra** [comment]
I guess it should have two modes: Either you start keepassx GUI and that provides the interface, or it is launched if you enable the autostart on activation service. Personally, I probably don't care if it starts the full GUI or just a one-shot minimal service. As long as the interface is already provided, it won't fire something up at all. At least this is how I guess it could work. It's probably more complicated in the background, I'm no dbus programmer.

**Comment by guillaume-uH57J9** [comment]
To be useful (ie and make the database accessible), I'm guessing keepassxc needs to be unlocked first, trough its GUI. 

So I don't the advantage in adding a service here. A service without GUI might be able to start earlier in the boot process (in the case of a system service) or earlier in the user session (in the case of a user daemon). But I see no point in doing that, even if the service can be started earlier than the GUI, it won't be able to serve requests until the GUI is started and the database unlocked.

**Comment by kakra** [comment]
Well, I wrote "one-shot minimal service" because I've thought of something that just asks for the unlock password, provides access, and then exits again. As it doesn't have the full-blown feature set (password generator, editing), it should be able to load fast and use less memory.

But memory is cheap these days, I don't really get the point why such an activation request shouldn't pull up the complete GUI daemon. I was just giving an example of a possible solution for people concerned for whatever reason with running the GUI process all the time. Personally, resource usage should not be a concern, especially thinking of opening and closing the databases and spawning processes every time an activation is made could be much more expensive.

If dbus activation works well enough, keepassxc could implement an option to completely shut the GUI down after a certain timeout (like 10 minutes or whatever feels right).

**Comment by guillaume-uH57J9** [comment]
That's a good point, sorry for missing your point on this service being one-shot and D-Bus activated. 

Not sure if keepassxc would be memory hungry. 

Measuring memory usage is not obvious. I ran a quick test wish massif which shows heap memory usage (malloc/new/...) is around 13.6M which sounds reasonable. 
However htop report a much higher memory usage of 832M, presumably because of shared libraries takes a lot of space. So it depends if you want to consider shared memory here. 

I'd already be very happy if keepassxc was available through libsecret/DBus, and would argue adding the option to have a one-short service is secondary.

**Comment by phoerious** [comment]
832MB is probably virtual memory, not actually consumed memory.

**Comment by mjsir911** [comment]
I'm not very well-versed with the KeepassXC codebase, but I am trying to keep extra dependencies needed for this issue at a minimum.

Does KeepassXC have everything needed for [this algorithm(s?)](https://standards.freedesktop.org/secret-service/ch07s03.html).
Just from a quick look I can't find that it uses a library that supports DH, aka Diffie-Hellman

And if I am forced to add an additional dependency, is that alright, any preferences for crypto libraries? I've very slightly dabbled with [crypto++](https://www.cryptopp.com/) and that seems to do what I need.

**Comment by droidmonkey** [comment]
Libsodium, already a dependency, probably has what you need.

https://download.libsodium.org/doc/advanced/scalar_multiplication.html

**Comment by Grief** [comment]
I am not sure, but I assume that kwallet (KDE) and gnome-keyring (Gnome) have significant differences, otherwise there won't be tons of unanswered questions ([like this one](https://askubuntu.com/questions/794039/replace-kwallet-with-gnome-seahorse)) about how to replace one with another in the internet.

**Comment by hrehfeld** [comment]
What's the status on this?

I can add the following user stories:

- I need to access my keepass data from offlineimap (python) via the python keyring library without additional password prompts

**Comment by droidmonkey** [comment]
I have considered starting a bounty run for KeePassXC. I looked into a couple services, they all looked a tad sketchy. Do you have a suggestion?

**Comment by virtualdxs** [comment]
https://www.bountysource.com/ is the one I've seen repeatedly

**Comment by intika** [comment]
Or simpler an issue with a paypal account

**Comment by virtualdxs** [comment]
That's less transparent and has less accountability. I wouldn't donate through that.

**Comment by alexander255** [comment]
Well, here we go: https://www.bountysource.com/issues/53971701-make-kpxc-databases-accessible-via-libsecret-dbus

I, as a developer of other stuff not KeePass*, just spent a little bit of time looking into this. Some observations:

  * Implementing the Secret Service D-Bus API isn't that hard, but likely somewhat time intensive:
      * The bus level crypto (for transferring secrets) is optional, so one can start without this for testing and tackle it later on when everything else works – a proper implementation will need this through!
  * It isn't immediately obvious how to map the API's support for “non-default collections”: In theory they'd be a perfect match for accessing different `.kdbx` database files over the bus, but how would one know which `.kdbx` file matches which collection alias names?
      * In practice this likely requires some kind of management application to be written which will ask users where new collections should be stored and remembers their mapping in a user-configurable way
      * Another problem would likely be the “single-instance” feature of KeePassXC which only allows one database to be open at a time (correct my if I'm wrong!)
      * An initial implementation could likely get away with just punting on this feature entirely and only supporting the default collection which basically all apps seem to use
   * Running the secret storage daemon as a separate application is not only possible, but also preferable in my opinion since it keeps the KeePassXC core smaller and results in a clear boundary between core and this platform integration ([Apple has a similar thing for instance](https://github.com/keepassxreboot/keepassxc/issues/2496), but see the next bullet however!)
      * Also doing it this way would allow for another cool feature: Unlocking the default KeePass database file using the login password (only GUI-less applications can do this and KeePassXC is not one of them)
   * There does not seem to be any documentation of KeePassXC's native protocol, which is bad (but there are [some client libraries](https://github.com/keepassxreboot/keepassxc/issues/2546) that should ease the pain and provide some ad-hoc documentation here)

If there is interest I'll try providing a time estimation (in hours) for throwing together a suitable proxy/bridge application that exposes the currently open database as default collection.

**Comment by kakra** [comment]
@alexander255 
> Well, here we go: https://www.bountysource.com/issues/53971701-make-kpxc-databases-accessible-via-libsecret-dbus

I'd prefer if the bounty would also mention KDE integration, not just Gnome...

**Comment by virtualdxs** [comment]
@kakra Only @kwill can change that. It's the issue text. Either way, they're not separate tasks.

**Comment by Skycoder42** [comment]
@kakra The SecretService API is generally independend of the Desktop Environment, as it is simply a standardized Specification by the Freedesktop foundation. Integration with GNOME/KDE specific applications has to be done by the developers of those and is seperated from the development of the service itself.

Also, I haven't seen that specification linked here (Might have overlooked it though). So here it is: https://specifications.freedesktop.org/secret-service/

**Comment by alexander255** [comment]
@kakra @Skycoder42:

> I'd prefer if the bounty would also mention KDE integration, not just Gnome...

I hate to be that guy, but one very important limitation regarding KDE here is that most “KDE-native” applications (most importantly the KDE-PIM/Kontact suite and KIO) still do not actually use the `libsecret`/SecretService API but rather KWallet which does not in any way interact with the secret storage API.

To highlight this, since I saw [the opposite](https://github.com/keepassxreboot/keepassxc/issues/1403#issuecomment-365399397) being claimed here as well:
**KWallet has nothing to do with `libsecret`, hence none of this discussion will change anything about it!**

Now, you may be saying “What about KSecretService?” Well, [KSecretService](https://community.kde.org/KSecretService) does not exist (that Wiki page is actually quite representative of the project in an ironical sense) except for an [early proof-of-concept from 2010](https://cgit.kde.org/scratch/vrusu/ksecretservice.git/). See also [this KDE bug](https://bugs.kde.org/show_bug.cgi?id=313216) for the current “status” – and yes that's me in that bug talking about dumping the project altogether.

Of course none of this precludes KDE project switching away from KWallet if there is a suitable alternative, but that's the status quo. The alternative would be to re-implement `libkf5wallet5` to use the KeePassXC interface instead, but that is definitely not in scope of this issue/project.

Sorry if this raised any false hopes!

**Comment by kabili207** [comment]
Thanks for pointing that out @alexander255. There's nothing wrong with being "that guy." I had based my comment off the freedesktop Secret Service page that listed the KDE Wallet devs among those who developed the API, but hadn't bothered to check if they had actually implemented it.

Considering the fact that the 0.2 spec hasn't had a meaningful update since **2011**, I highly doubt KWallet will be updated to support the Secret Service API anytime in the foreseeable future, so I'd agree that anyone seeking compatibility should submit an issue requesting support.

**Comment by intika** [comment]
An other bounty for KDE ? lol 

I guess a good implementation would be an upstreamed KDE compatibility with the keepass system (in kde wallets management) `libkf5wallet5` + `kwalletmanager5` but this would be even more work than libsecret... 

Any way KDE does not need Keepassxc as much as Gnome Keyring 

One important thing while developing this feature **if possible** is keyring isolation which does not exist on the current gnome keyring (one application requesting credentials can access all the content) may be after the initial release... just my 2 cents 

;)

**Comment by alexander255** [comment]
> One important thing while developing this feature **if possible** is keyring isolation which does not exist on the current gnome keyring (one application requesting credentials can access all the content) may be after the initial release... just my 2 cents

I've just spent the last 1,5 hours trying to define what **app isolation** would even mean in this context. :slightly_smiling_face:
A TL;DR may be in order:

  * We can only securely distinguish running binaries from one-another, script-based applications will be reduced to their respective scripting language interpreter and any further subdivision would be based on the respective application author's good will
  * Sandboxed applications (Flatpak) are not harder than distinguishing binaries from one-another and should reliably work no matter what they contain

In this context isolation means: “If a new application tries to use an existing password a prompt will be displayed” – All applications still have access to everything with user consent (there are no separated namespaces).

**Comment by kakra** [comment]
@alexander255 Please don't hate being that guy. This is valuable information. The pure existence of that said wiki page adds to the confusion, including the fact that my system had once this API bridge from the proof-of-concept project installed. That explains something, as I was wondering lately why Gnome apps do not see my wallet passwords.

So in the end it's "that topic again": One other project of KDE silently vanished as not being continued.

@intika If there should be one bounty, maybe one that revived that old proof-of-concept project and turns it into a first class citizen? I'm not sure why this project has been suspended. Did it encounter major incompatibilities with the KDE ecosystem? Was it just too early and the secrets API hasn't developed enough yet?

**Comment by kabili207** [comment]
@kakra The original dev who worked on it stopped [because of kdelibs splitting into KF5](https://community.kde.org/Akademy/2015/AllBoF/KSecret_Service). The person who re-started work on it in 2015 dropped it about a year later because [he no longer had the time](https://marc.info/?l=kde-devel&m=146788782429932).

**Comment by alexander255** [comment]
@kara Thanks! Do, by bridge, mean something that exposes a SecretService implementation and forwards those requests to KWallet? If so, reviving it won't do any good as it be forwarding stuff in the wrong direction: You need something that translates from *KWallet* to *SecretStorage*, not the other way round.

Also from the discussion @kabili207 linked it appears like KWallet is actually a D-Bus API as well? At least some quick tests seem to confirm this. Also [this](https://code.woboq.org/qt5/kf5/kwallet/src/api/KWallet/kwallet.cpp.html) appears to be the entire client library and one see an incomplete implementation of a SecretStorage based backend there. Both, completing and activating this backend, or reimplementing the D-Bus API of course remain issues not directly related to this one, but they highlight possible ways forward.

**Comment by alexander255** [comment]
I've pushed an estimate on the price of this project to BountyHunters:
22 hours of work of which 4 hours are the required research (already done, [see my first post](https://github.com/keepassxreboot/keepassxc/issues/1403#issuecomment-457949605)) and 18 hours (or 3 days @ 6 hours a day) for implementation at a wage of 10$/h.

The implementation will cover: 

  * The main mediator service two-way translating `SecretStorage` API calls to the KeePassXC browser protocol including the highly recommended bus-level encryption

The implementation may cover:

  * Client-level isolation – if it doesn't only the key provisioning and app identity detection will be missing making it easy to fix this later on
  * Automatically unlocking KeePassXC with the login password – probably won't happen within the budgeted timeframe, but I may add it anyways 'cause it's a cool feature

The implementation will definitely not cover:

  * Non-default collections (which would effectively expose multiple databases over the bus at the same time)

**Comment by virtualdxs** [comment]
@alexander255 Regarding Automatically unlocking KeePassXC with the login password: I love this idea. I will say though, it seems like it deserves its own issue.

**Comment by alexander255** [comment]
Yes, probably. But the way to do this afaik is to spawn off a service during PAM's session phase that will then store that password in memory and forward it to KeePassXC once that has been started. That service could IMHO be the proposed SecretStorage bridge, since it's basically what will do “Linux integration” anyways.

**Comment by virtualdxs** [comment]
I see, that makes sense then to keep it in this issue if that's how it's to be implemented.

**Comment by jo-so** [comment]
> The implementation will cover: 
> 
>   * The main mediator service two-way translating `SecretStorage` API calls to the KeePassXC browser protocol including the highly recommended bus-level encryption

Why we need a new service? Couldn't be the SecretStorage API implemented
in KeePassXC itself? Why we need to go the way through the browser
protocol? What lacks the SecretStorage API that makes it necessary to got
through the browser API?

**Comment by alexander255** [comment]
@jo-so:

> Why we need a new service? Couldn't be the SecretStorage API implemented in KeePassXC itself?

We don't *need* a new service and of course all of this can be implemented as part of KeePassXC core itself. My rationale for going implementation is this:

> Running the secret storage daemon as a separate application is not only possible, but also preferable in my opinion since it keeps the KeePassXC core smaller and results in a clear boundary between core and this platform integration (Apple has a similar thing through a totally different API for instance)

Please outline why you think it's preferable to stuff it all in one big binary. I'm genuinely curios. :slightly_smiling_face: 

> Why we need to go the way through the browser protocol?

Because it's the **only** supported protocol (with KeePassHTTP and KeePassRPC being deprecated/dropped)?

> What lacks the SecretStorage API that makes it necessary to got through the browser API?

Nothing, they map pretty well actually. The browser protocol is better insofar as that it forces clients to provision their own HOTP key for encryption and authentication rather than regressing to a basic “if it can connect, it's trusted for everything”. See point 4 of my [proposal/notes](https://github.com/keepassxreboot/keepassxc/issues/1403#issuecomment-457949605) on my ideas on how this can be implemented (within certain limits) based on the sender's identity however.

I will admit however that provisioning those keys inside the bridge however is not ideal, but it's not like we can upgrade the *existing* SecretStorage protocol with this property.

**Comment by nazar-pc** [comment]
Yes, Nextcloud and I believe IntelliJ IDEA uses it as well, I'd be happy to remove Gnome's implementation from my system entirely.

**Comment by affinityv** [comment]
Please limit any requirements to require dbus and systemd -- they should be optional only as there are alternatives and a great number of people whom understand that systemd is a cancer on Linux.

**Comment by virtualdxs** [comment]
@affinityv I agree that there should not be a dependency on systemd. That being said, it's inappropriate to call systemd a "cancer", as it's both an inaccurate analogy and disrespectful to those who have lost someone to cancer.

**Comment by MurzNN** [comment]
For make KPXC available via DBus we can implement, at first step, access to passwords via dbus commands, like KWallet do now:
```
method QByteArray org.kde.KWallet.readEntry(int handle, QString folder, QString key, QString appid)
method QVariantMap org.kde.KWallet.readEntryList(int handle, QString folder, QString key, QString appid)
method QByteArray org.kde.KWallet.readMap(int handle, QString folder, QString key, QString appid)
method QVariantMap org.kde.KWallet.readMapList(int handle, QString folder, QString key, QString appid)
method QString org.kde.KWallet.readPassword(int handle, QString folder, QString key, QString appid)
method QVariantMap org.kde.KWallet.readPasswordList(int handle, QString folder, QString key, QString appid)
method int org.kde.KWallet.removeEntry(int handle, QString folder, QString key, QString appid)
method bool org.kde.KWallet.removeFolder(int handle, QString folder, QString appid)
method int org.kde.KWallet.renameEntry(int handle, QString folder, QString oldName, QString newName, QString appid)
method int org.kde.KWallet.writeEntry(int handle, QString folder, QString key, QByteArray value, QString appid)
method int org.kde.KWallet.writeEntry(int handle, QString folder, QString key, QByteArray value, int entryType, QString appid)
method int org.kde.KWallet.writeMap(int handle, QString folder, QString key, QByteArray value, QString appid)
method int org.kde.KWallet.writePassword(int handle, QString folder, QString key, QString value, QString appid)
```

**Comment by alexander255** [comment]
@virtualdxs: `systemd` is only relevant insofar as that you can have D-Bus autostart service (autoactivation) on first access and that autoactivation *can* defer to a systemd service. I wasn't really planning on providing the relevant activation files, but writing (read: copying an existing and adapting) one is trivial and if people want this I see no reason why it shouldn't work.

@MurzNN: This proposal is specifically about implementing a SecretService-compatible service. Exposing KWallet or a custom API is not planned at this time. However exposing KWallet once SecretService is in place will be able to reuse most code and hence will only require modest extra effort.
As for exposing a custom API: Honestly I don't see any good reason why “native client” should use D-Bus rather then just connecting to the relevant Unix socket directly. Just for the sake of completeness maybe? Also please don't take any inspiration from KWallet however: Their API is the worst in a lot of ways (cleartext passwords, useless client isolation) and a new protocol should definitely **not** attempt to replicate this.

**Comment by virtualdxs** [comment]
@alexander255 Just to clarify, I'm specifically referring to the message `** Message: 22:22:32.970: Remote error from secret service: org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.secrets was not provided by any .service files`. If that's just about activation, then that's perfect. I'm very familiar with the systemd unit file syntax.

**Comment by alexander255** [comment]
@virtualdxs: That message just is D-Bus daemon saying: *Nobody provides the SecretService API and I can't figure out how to fix it*. If you'd start an application providing the API manually the message would go away as well – activation is only attempted since the API wasn't provided by anybody, it then fails however causing the error message you observe.

**Comment by hrehfeld** [comment]
Awesome! :+1:

**Comment by abeluck** [comment]
Any ideas when this will make it into a release?

**Comment by droidmonkey** [comment]
It will be in the 2.5.0 release.

**Comment by guillaume-uH57J9** [comment]
Thanks for doing this. I'm too late for the chip in for the bounty, so I setup a recurring contribution using liberapay
